federation/pkg/federation-controller/service/dns.go

```diff
@@ -263,6 +266,9 @@ func (s *ServiceController) ensureDnsRecords(clusterName string, cachedService *
 	serviceName := cachedService.lastState.Name
 	namespaceName := cachedService.lastState.Namespace
 	zoneNames, regionName, err := s.getClusterZoneNames(clusterName)
+	if zoneNames == nil {
+		return fmt.Errorf("fail to get cluster zone names")
```

Comment:
Oh, and either way, the correct approach would be to check err before checking zoneName == nil.  The way the code is now, you will completely discard all errors returned by getClusterZoneNames (because whenever a non-nil error is returned, zoneNames will also be nil.